added embrace campaign model and exposed it via graphql.

nikita-pardeshi-github opened the pull request:
Implemented [Groot-150](https://linear.app/bankgreen/issue/GROOT-150/integrate-embrace-campaign-model-into-bankgreen-django-and-expose-via)

Development steps 

1. Added embrace campaign model
2. Integrated embrace campaign model with commentary model (many to many relationship)
3. Added embrace campaign in admin interface
4. Exposed this model data via GraphQL

RogerTangos commented on brand/migrations/0041_embracecampaign.py:
```diff
@@ -0,0 +1,25 @@
+# Generated by Django 4.1.7 on 2024-03-26 18:22
+
+from django.db import migrations, models
```
When putting together a commit with migrations, it's best to squash your commits. The easiest way to achieve this is to delete the migration files locally and `python manage.py make migrations` another time. Then they'll all be in one file

RogerTangos commented on GraphQL-README.md:
```diff
@@ -29,6 +29,7 @@ query BrandsQuery(
         commentary @include(if: $withCommentary) {
           rating
           embrace
+          embrace
```
typo

RogerTangos replied:
I tried creating an embrace campaign and querying a bank with it enabled. It returned "none" though, which doesn't seem right

nikita-pardeshi-github replied:
I tried add embrace campaign in existing brand and with creating new brand. It worked for me. Let me know if I am missing anything here.Below are the results of graphql query

query BrandByNameQuery($name: String!) {
  brandByName(name: $name) {
    tag
    commentary {
      embraceCampaign {
        id
      }
    }
  } 
}
  output : {
  "data": {
    "brandByName": {
      "tag": "hiver",
      "commentary": {
        "embraceCampaign": [
          {
            "id": "18"
          }
        ]
      }
    }
  }
}
------------------
{
  "data": {
    "brandByName": {
      "tag": "test_embrace_campaign",
      "commentary": {
        "embraceCampaign": [
          {
            "id": "18"
          }
        ]
      }
    }
  }
}

RogerTangos commented on brand/admin.py:
```diff
@@ -314,3 +316,15 @@ def changelist_view(self, request, extra_context=None):
+
+    model = EmbraceCampaign
+
+    list_display = ["name", "description"]
```
You'll probably want to display id here, for the developers to see. That's what they'll be using when configuring new campaigns

RogerTangos commented on brand/models/commentary.py:
```diff
@@ -7,6 +7,7 @@
 from django_countries.fields import CountryField
```
It looks like CountryField and Enum are unused. I has my git blame on it, but don't hesitant to remove stuff like that as long as the tests still pass. You can check to make sure no one is importing this in a roundabout way by searching `from Commentary import CountryField`. If there aren't matches then you're probably good to remove it

RogerTangos commented on brand/models/embrace_campaign.py:
```diff
@@ -0,0 +1,10 @@
+from django.db import models
+
+
+class EmbraceCampaign(models.Model):
```
This certainly needs a validator to make sure that the configuration JSON is how you expect it to be. Checkout how `Brand` calls a function, `validate_tag`

What fields are required? "System Prompt" and "User Prompt" I'm assuming. Anything else?

RogerTangos commented on brand/schema.py:
```diff
@@ -259,12 +267,18 @@ class Query(graphene.ObjectType):
         return BrandModel.objects.get(name=name)
 
+    def resolve_embrace_campaigns(root, info):
+        print(EmbraceCampaignModel.objects.all())
```
extra print statement

RogerTangos commented:
A last thing - if the user is expected to type JSON in the admin panel, it would be helpful to supply them with a JSON editor. This isn't strictly necessary but if it's easy then it's worth doing

nikita-pardeshi-github commented:
@RogerTangos I have implemented all the review comments. Could you please review it?

RogerTangos commented on brand/models/embrace_campaign.py:
```diff
@@ -0,0 +1,38 @@
+def validate_configuration(json_config):
+    schema_keys = ["email", "system_prompt", "user_prompt"]
+    schema_default_values = [
+        "write email address here",
```
It would certainly be better to have a default address filled in

nikita-pardeshi-github commented on brand/models/embrace_campaign.py:
```diff
@@ -0,0 +1,38 @@
+
+    unknown_keys = set(json_config.keys()) - set(schema_keys)
+    if unknown_keys:
+        raise ValidationError(
```
Thanks @RogerTangos  for reviewing it. Yeah, I will remove strict validation of the keys so that we can add additional keys later if required.

RogerTangos replied:
thanks!

nikita-pardeshi-github commented:
@RogerTangos I have modified validate_configuration method with jsonschema. Also added default email address. Could you please confirm once if this is okay?

RogerTangos commented on brand/models/embrace_campaign.py:
```diff
@@ -0,0 +1,36 @@
+
+    try:
+        validate_email(json_config["email"])
+    except:
```
It's bad practice to bare except statement here. If there's an exception that's typically thrown then just test for that

RogerTangos commented:
Merged and squashed. I noticed another issue and made a comment on it, but think that you should just fix it in a next PR. It's not a big deal here
